test/lorenz.jl

```diff
@@ -0,0 +1,222 @@
+end
+
+@testset "Lorenz system" begin
+    tspan=(0.0, 0.5)
+    pvec = [2.0, 1.0, 8.0/3.0]
+    x0 = [1.0, 1.0, 1.0]
+    nsteps = 1000
+    nx, np = length(x0), length(pvec)
+    ϵ = 1e-7
+
+    lc = LorenzContext()
+
+    # system dynamics
+    sys = CLSDyn.SystemDynamics(lc, pvec, lorenz_rhs!)
+    problem = CLSDyn.IVP(nsteps, x0, tspan, CLSDyn.RK(), sys)
+    CLSDyn.set_fx!(sys, lorenz_fx!)
+    CLSDyn.set_fp!(sys, lorenz_fp!)
+    CLSDyn.set_fx_trans!(sys, lorenz_fx_trans!)
+    CLSDyn.set_fp_trans!(sys, lorenz_fp_trans!)
+    CLSDyn.set_fxx!(sys, lorenz_fxx!)
+    CLSDyn.set_fxp!(sys, lorenz_fxp!)
+    CLSDyn.set_fpx!(sys, lorenz_fpx!)
+    CLSDyn.set_fpp!(sys, lorenz_fpp!)
+
+    # costs
+    cost = CLSDyn.CostFunctional(sys, r, nothing)
+    CLSDyn.set_rx!(cost, rx)
+    CLSDyn.set_rp!(cost, rp)
+    CLSDyn.set_rxx!(cost, rxx)
+    CLSDyn.set_rpp!(cost, rpp)
+    CLSDyn.set_rxp!(cost, rxp)
+    CLSDyn.set_rpx!(cost, rpx)
+
+    # Utils
+    function numerical_integration(x0, pvec)
+        traj, tvec = CLSDyn.rk4(sys.rhs!, x0, pvec, tspan, nsteps)
+        val = 0.0
+        for i=1:(nsteps-1)
+            val += (tvec[i + 1] - tvec[i])*r(traj[:, i + 1], pvec, i, lc)
+        end
+        return val
+    end
+
+    function terminal_condition(x0)
+        traj, tvec = CLSDyn.rk4(sys.rhs!, x0, pvec, tspan, nsteps)
+        return traj[1,end]
+    end
+
+    fint_x(x0) = numerical_integration(x0, pvec)
+    fint_p(pvec) = numerical_integration(x0, pvec)
+
+    # Trajectory
+    traj, tvec = CLSDyn.solve(problem)
+    x0eps = [x0[1] + ϵ, x0[2], x0[3]]
+
+    problem2 = CLSDyn.IVP(nsteps, x0eps, tspan, CLSDyn.RK(), sys)
+    traj2, tvec2 = CLSDyn.solve(problem2)
+    @test traj ≈ traj2 atol=1e-5
+    @test size(traj) == (nx, nsteps)
+
+    @testset "Forward sensitivities" begin
+        dx0 = [1.0, 0.0, 0.0]
+        tlm_traj = CLSDyn.tlm(problem, traj, tvec, dx0)
+        @test size(tlm_traj) == (np, nsteps)
+    end
+
+    λ0 = [0.0, 0.0, 0.0]
+    @testset "Adjoint sensitivities" begin
+        λ, μ = CLSDyn.adjoint_sens(problem, cost, traj, tvec, λ0)
+        @test size(λ) == (nx, )
+        @test size(μ) == (np, )
+
+        λ_fd = FiniteDiff.finite_difference_gradient(fint_x, x0)
+        @test λ ≈ λ_fd atol=1e-4
+
+        μ_fd = FiniteDiff.finite_difference_gradient(fint_p, pvec)
+        @test μ ≈ μ_fd atol=1e-4
+    end
+
+    @testset "Second-order sensitivities" begin
+        λ_traj = similar(traj)
+        CLSDyn.adjoint_sens(problem, cost, traj, tvec, λ0; λ_traj=λ_traj)
+        v = [0.0, 0.0, 0.0]
+        δθ = [1.0, 0.0, 0.0]
+        σ, τ = CLSDyn.forward_over_adjoint(
+            problem,
+            cost,
+            traj,
+            λ_traj,
+            tvec,
+            pvec,
+            v,
+            δθ
+        )
+        @test size(σ) == (nx, )
+        @test size(τ) == (np, )
+        H = FiniteDiff.finite_difference_hessian(fint_x, x0)
+        @test_broken σ ≈ H * δθ atol=1e-4
```

Comment:
@dmaldona I am not sure how to fix this ...

Reply:
@frapac This one is tricky

If you want sensitivities of the functional w.r.t. the first parameter, then set
`        v = [0.0, 0.0, 0.0]
        δθ = [1.0, 0.0, 0.0]`
        
However, if you want sensitivities of the functional w.r.t the first initial condition, then set
`        v = [1.0, 0.0, 0.0]
        δθ = [0.0, 0.0, 0.0]`
        
This should give you the correct sensitivities compared with finite differences.

Now, this raises a couple question: In the first-order adjoint, lambda is the sensitivity of the functional w.r.t the initial condition and nu is w.r.t the parameter, and we get it on one shot.

To be honest, this is why I expected and I am not sure of why is the case. I need to look closer at the equations and make sure I did not make a mistake.

But for now, I believe that the code above should work.